### PR TITLE
#430 - TkRetry and TkRetry test

### PR DESCRIPTION
**Pull request by aschworer**
#430 TkRetry and TkRetry test

### CODE REVIEW

**Comment by aschworer** [comment]
@davvd @dmzaytsev #430 is ready for code review here

**Comment by davvd** [comment]
@aschworer Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@dmzaytsev review this one,please

File: src/main/java/org/takes/tk/TkRetry.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import org.takes.Request;
+import org.takes.Response;
+import org.takes.Take;
+
+/**
+ * Decorator TkRetry, which will not fail immediately on IOException, but
+ * will retry a few times.
+ *
+ * @author Aygul Schworer (agul.schworer@gmail.com)
+ *
```
Comment (dmzaytsev) [inline src/main/java/org/takes/tk/TkRetry.java]: @aschworer I think you can remove empty line here

File: src/main/java/org/takes/tk/TkRetry.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * @version $Id$
+ *
```
Comment (dmzaytsev) [inline src/main/java/org/takes/tk/TkRetry.java]: @aschworer `@since 0.28.3`

File: src/test/java/org/takes/tk/TkRetryTest.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import org.takes.rq.RqFake;
+import org.takes.rq.RqMethod;
+import org.takes.rs.RsPrint;
+
+/**
+ * Test for TkRetry.
+ *
+ * @author Aygul Schworer (aygul.schworer@gmail.com)
+ * @version $Id$
+ *
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkRetryTest.java]: @aschworer @since 0.28.3

File: src/test/java/org/takes/tk/TkRetryTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+ */
+public final class TkRetryTest {
+
+    /**
+     * TkRetry works.
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkRetryTest.java]: @aschworer please check out this article http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names
and update javadoc and method names here and below

File: src/test/java/org/takes/tk/TkRetryTest.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        final int delay = 1000;
+        final Take take = Mockito.mock(Take.class);
+        final Request req = new RqFake(RqMethod.GET);
+        Mockito.when(take.act(req)).thenThrow(new IOException());
+        final long minTimeToFail = count * delay;
+        final long startTime = System.currentTimeMillis();
+        try {
+            new TkRetry(count, delay, take).act(req);
+        } catch (final IOException exception) {
+            final long stopTime = System.currentTimeMillis();
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkRetryTest.java]: @aschworer once used variable can be inline

File: src/test/java/org/takes/tk/TkRetryTest.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            final long elapsedTime = stopTime - startTime;
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkRetryTest.java]: @aschworer same here

**Comment by dmzaytsev** [comment]
@aschworer please see a few comments above
Also I think we need one more test
In this test method #act() of the origin take should raise an exception 	
initially and return without any exception on the next call

**Comment by dmzaytsev** [comment]
@aschworer ping

**Comment by aschworer** [comment]
@dmzaytsev Thanks Dmitry for the comments. I'm still new to here, it's my first task, so I'm taking some time to read stuff and get familiar. Plus its Christmas here. I'll do my best to fix the review in the next 2-3 days.

**Comment by dmzaytsev** [comment]
@aschworer no problem at all
happy holidays!

**Comment by aschworer** [comment]
@dmzaytsev the fixes are done. could you please check?

File: src/test/java/org/takes/tk/TkRetryTest.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+ * TkRetry can retry till success or retry count is reached.
+ *
+ * @author Aygul Schworer (aygul.schworer@gmail.com)
+ * @version $Id 7601ff45d71fde024bac703043555d83c08269e7 $
+ * @since 0.28.3
+ */
+public final class TkRetryTest {
+
+    /**
+     * TkRetry can work when no no IOException.
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkRetryTest.java]: @aschworer typo double `no`

File: src/test/java/org/takes/tk/TkRetryTest.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+     *
+     * @throws Exception if something is wrong
+     */
+    @Test
+    public void justWorks() throws Exception {
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkRetryTest.java]: @aschworer `worksWhenNoException`
www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names

File: src/test/java/org/takes/tk/TkRetryTest.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    @Test(expected = IOException.class)
+    public void retriesOnExceptionTillCount() throws Exception {
+        final int count = 3;
+        final int delay = 1000;
+        final Take take = Mockito.mock(Take.class);
+        final Request req = new RqFake(RqMethod.GET);
+        Mockito.when(take.act(Mockito.any(Request.class))).thenThrow(
+                new IOException()
+        );
+        final long minTimeToFail = count * delay;
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkRetryTest.java]: @aschworer why we need this variable? it used only once

File: src/test/java/org/takes/tk/TkRetryTest.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        final int delay = 1000;
+        final Take take = Mockito.mock(Take.class);
+        Mockito.when(take.act(Mockito.any(Request.class))).thenThrow(
+                new IOException()
+        ).thenReturn(new RsText());
+        final long minTime = delay;
+        final long startTime = System.currentTimeMillis();
+        new TkRetry(count, delay, take).act(new RqFake(RqMethod.GET));
+        MatcherAssert.assertThat(
+            minTime,
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkRetryTest.java]: @aschworer same here, once used variable `minTime` can be replaced by `delay`

**Comment by dmzaytsev** [comment]
@aschworer please see a few minor comments above

**Comment by aschworer** [comment]
@dmzaytsev thanks for the comments. all done

**Comment by dmzaytsev** [comment]
@aschworer thanks

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by dmzaytsev** [comment]
@yegor256 ping

**Comment by yegor256** [comment]
@aschworer I believe this pull request is no longer important, since #430 is re-assigned and new solution is proposed in #493. I'm closing it

**Comment by davvd** [comment]
@ypshenychka please, review this task for compliance with our [quality rules](http://at.teamed.io/qa.html)

**Comment by ypshenychka** [comment]
@davvd Quality is good.

**Comment by davvd** [comment]
> @davvd Quality is good.

@ypshenychka thank you :)

**Comment by davvd** [comment]
@dmzaytsev I added 10 mins to @ypshenychka (for QA review) in transaction 73944272... thanks, paid, **32 mins** to your account, payment ID is `73944292`, task took 390 hours and 43 mins to complete... review comments (c=17) added as a bonus... added +32 to your rating, now it is equal to [+1257](http://www.netbout.com/b/35166?open=rating)

**Pull request closed without merging**